pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+    common::{DispatchOutcome, JournalNote},
+    process_non_executable, SystemReservationContext,
+};
+use gear_core::{
+    ids::{BuiltinId, MessageId, ProgramId},
+    message::{DispatchKind, ReplyMessage, ReplyPacket, StoredDispatch},
+};
+use gear_core_errors::SimpleExecutionError;
+use impl_trait_for_tuples::impl_for_tuples;
+use pallet_gear::{BuiltinRouter, BuiltinRouterProvider};
+use parity_scale_codec::{Decode, Encode};
+use sp_io::hashing::blake2_256;
+use sp_runtime::traits::{TrailingZeroInput, Zero};
+use sp_std::prelude::*;
+
+pub use pallet::*;
+
+#[allow(dead_code)]
+const LOG_TARGET: &str = "gear::builtin_actor";
+
+pub trait Dispatchable {
+    type Payload: AsRef<[u8]>;
+
+    fn source(&self) -> ProgramId;
+    fn destination(&self) -> BuiltinId;
+    fn payload_bytes(&self) -> &[u8];
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct SimpleBuiltinMessage {
+    source: ProgramId,
+    destination: BuiltinId,
+    payload: Vec<u8>,
+}
+
+impl Dispatchable for SimpleBuiltinMessage {
+    type Payload = Vec<u8>;
+
+    fn source(&self) -> ProgramId {
+        self.source
+    }
+
+    fn destination(&self) -> BuiltinId {
+        self.destination
+    }
+
+    fn payload_bytes(&self) -> &[u8] {
+        self.payload.as_ref()
+    }
+}
+
+/// Built-in actor error type
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::Display)]
+pub enum BuiltinActorError {
+    /// Occurs if the underlying call has the weight greater than the `gas_limit`.
+    #[display(fmt = "Not enough gas supplied")]
+    InsufficientGas,
+    /// Occurs if the dispatch's message can't be decoded into a known type.
+    #[display(fmt = "Failure to decode message")]
+    UnknownMessageType,
+    /// Occurs if a builtin id doesn't belong to any registered actor.
+    #[display(fmt = "Unknown builtin id")]
+    UnknownBuiltinId,
+}
+
+// TODO: adjust according to potential changes in `SimpleExecutionError` that might be coming
+impl From<BuiltinActorError> for SimpleExecutionError {
+    /// Convert [`BuiltinActorError`] into [`gear_core_errors::SimpleExecutionError`].
+    fn from(err: BuiltinActorError) -> Self {
+        match err {
+            BuiltinActorError::InsufficientGas => SimpleExecutionError::RanOutOfGas,
+            BuiltinActorError::UnknownMessageType => SimpleExecutionError::UserspacePanic,
+            BuiltinActorError::UnknownBuiltinId => SimpleExecutionError::BackendError,
+        }
+    }
+}
+
+pub type BuiltinResult<P> = Result<P, BuiltinActorError>;
+
+/// A trait representing an interface of a builtin actor that can receive a message
+/// and produce a set of outputs that can then be converted into a reply message.
+pub trait BuiltinActor<Message: Dispatchable, Gas: Zero> {
+    /// Handles a message and returns a result and the actual gas spent.
+    fn handle(message: &Message, gas_limit: Gas) -> (BuiltinResult<Message::Payload>, Gas);
+
+    fn get_ids(buffer: &mut Vec<BuiltinId>);
```

Comment:
fn get_ids() -> BTreeSet<>?

Reply:
not sure we can wrap this one inside `impl_for_tuples`. That's the reason of having it `C-style`. But I'll check.